### ISSUE TITLE
关于新加评价指标的实现代码

### ISSUE THREAD

**Issue by sound-and-fury** [HaitaoWuTJU/Uncertainty-aware-Blur-Prior#15]
感谢您的开源和一直以来的积极回复，我们注意到您在文章中写道Notably, we employed two additional evaluation metrics, mAP and Similarity Score,to comprehensively assess performance, as detailed in Appendix B.5.

<img width="901" height="672" alt="Image" src="https://github.com/user-attachments/assets/af470ea6-64b0-4023-aa17-f87736517781" />
我们现在也希望用mAP，Similarity Score指标评测我们的工作，但并未在您的仓库中找到相关的分析程序，请问您是否还留有当时的代码，如果方便的话，期待您的开源

**Comment by HaitaoWuTJU** [comment]
您好,非常感谢您的关注与认可,也感谢您细致地阅读了我们论文中的这一部分评估!
关于您提到的 mAP 和 Similarity Score 两个指标,相关实现其实已经包含在main.py 的 test_step 和 on_test_epoch_end 里),只是没有单独拆成一个评测脚本，所以不太容易一眼找到。这里简要说明一下实现方式，方便您复用:
- Similarity Score
对测试集中每一个 (EEG/MEG, 配对图像) 样本对,计算二者在归一化特征空间中的余弦相似度(即相似度矩阵的对角线元素),最后对整个测试集取平均
```python
# test_step 中
eeg_z = eeg_z / eeg_z.norm(dim=-1, keepdim=True)
similarity = eeg_z @ img_z.T                       # [B, B]
self.match_similarities.extend(similarity.diag().detach().cpu().tolist())

# on_test_epoch_end 中
self.match_similarities = np.mean(self.match_similarities)
self.log('similarity', self.match_similarities, sync_dist=True)
```

- mAP
由于每个 query(EEG/MEG)在检索池中只有 1 个 正确匹配图像,AP 退化为 1/rank(rank 为正确图像在按相似度降序排列后的位次),最后对所有样本取平均得到 mAP

```python
# test_step 中
for i in range(similarity.shape[0]):
    true_index = i
    sims = similarity[i, :]
    sorted_indices = torch.argsort(-sims)
    rank = (sorted_indices == true_index).nonzero()[0][0] + 1
    ap = 1 / rank
    self.mAP_total += ap

# on_test_epoch_end 中
self.mAP = (self.mAP_total / len(all_true_labels)).item()
self.log('mAP', self.mAP, sync_dist=True)
```
如果还有任何不清楚的地方，欢迎继续在 issue 下讨论。祝科研顺利!

**Comment by sound-and-fury** [comment]
好的，谢谢您的热心帮助！

**Comment by sound-and-fury** [comment]
您好，您在文章中提到As shown in Tab. 4, we report the Pearson and Spearman correlations between subject variability and zero-shot retrieval accuracy. Methods such as NICE-SA, VE-SDN, and Vanilla exhibit strong negative Pearson correlations (e.g., -0.783, -0.687, and -0.761 for top-1, respectively), indicating substantial sensitivity to subject variability. In contrast, UBP demonstrates improved robustness, with a less negative Pearson correlation (-0.481) compared to Vanilla. UBP demonstrates more stable performance when handling subjects with high variability, with its accuracy not significantly degrading.

<img width="711" height="545" alt="Image" src="https://github.com/user-attachments/assets/26ec9ef1-0782-4258-bca8-8423f508636d" />
我已经努力尝试过在仓库中寻找相关代码，不过我还是没能找到。不知道您是否还留有当时的代码，如果方便的话，期待您的开源，谢谢您为社区做出的巨大贡献！

**Comment by HaitaoWuTJU** [comment]
感谢您对这一部分的关注，这部分非核心实验的评估代码确实没有放在repo里面，核心思路是计算10个受试者的准确度和10个受试者的**variablity**，其两者的pearson，spearmanr相关性。variablity 是根据每个受试者对于同一视觉图片产生的4次trial之间的mse计算的variablity。原始代码可能需要一段时间找一下，这里仅给出参考：
```python
import numpy as np
from scipy.stats import pearsonr, spearmanr
from itertools import combinations


def mse(x, y):
    """Compute Mean Squared Error between two vectors/images."""
    return np.mean((x - y) ** 2)


def compute_subject_variability(subject_trials):
    """
    Compute variability for one subject.

    Args:
        subject_trials: dict
            {
                image_id_1: [trial1_feat, trial2_feat, trial3_feat],
                image_id_2: [trial1_feat, trial2_feat, trial3_feat],
                ...
            }

    Returns:
        float: averaged variability over all images
    """
    image_variabilities = []

    for image_id, trials in subject_trials.items():
        assert len(trials) == 3, f"{image_id} does not have 3 trials"

        pairwise_mse = []
        for a, b in combinations(trials, 2):
            pairwise_mse.append(mse(a, b))

        image_variabilities.append(np.mean(pairwise_mse))

    return np.mean(image_variabilities)


def compute_correlations(subject_accuracies, subject_variabilities):
    """
    Compute Pearson and Spearman correlation.

    Args:
        subject_accuracies: list[float]
        subject_variabilities: list[float]

    Returns:
        dict
    """
    pearson_corr, pearson_p = pearsonr(subject_accuracies, subject_variabilities)
    spearman_corr, spearman_p = spearmanr(subject_accuracies, subject_variabilities)

    return {
        "pearson_r": pearson_corr,
        "pearson_p": pearson_p,
        "spearman_r": spearman_corr,
        "spearman_p": spearman_p,
    }


# =========================
# Example Usage
# =========================

num_subjects = 10

subject_accuracies = []
subject_variabilities = []

for subject_id in range(num_subjects):

    # -------------------------
    # Replace this with your real accuracy
    # -------------------------
    accuracy = np.random.rand()
    subject_accuracies.append(accuracy)

    # -------------------------
    # Replace this with your real trial data
    # shape example: feature vector of dim 768
    # -------------------------
    subject_trials = {
        f"img_{i}": [
            np.random.randn(768),
            np.random.randn(768),
            np.random.randn(768),
        ]
        for i in range(100)
    }

    variability = compute_subject_variability(subject_trials)
    subject_variabilities.append(variability)


results = compute_correlations(subject_accuracies, subject_variabilities)

print("Subject Accuracies:", subject_accuracies)
print("Subject Variabilities:", subject_variabilities)
print("\nCorrelation Results:")
for k, v in results.items():
    print(f"{k}: {v:.6f}")

```

**Comment by sound-and-fury** [comment]
> 感谢您对这一部分的关注，这部分非核心实验的评估代码确实没有放在repo里面，核心思路是计算10个受试者的准确度和10个受试者的**variablity**，其两者的pearson，spearmanr相关性。variablity 是根据每个受试者对于同一视觉图片产生的3次trial之间的mse计算的variablity。原始代码可能需要一段时间找一下，这里仅给出参考：
> 
> import numpy as np
> from scipy.stats import pearsonr, spearmanr
> from itertools import combinations
> 
> 
> def mse(x, y):
>     """Compute Mean Squared Error between two vectors/images."""
>     return np.mean((x - y) ** 2)
> 
> 
> def compute_subject_variability(subject_trials):
>     """
>     Compute variability for one subject.
> 
>     Args:
>         subject_trials: dict
>             {
>                 image_id_1: [trial1_feat, trial2_feat, trial3_feat],
>                 image_id_2: [trial1_feat, trial2_feat, trial3_feat],
>                 ...
>             }
> 
>     Returns:
>         float: averaged variability over all images
>     """
>     image_variabilities = []
> 
>     for image_id, trials in subject_trials.items():
>         assert len(trials) == 3, f"{image_id} does not have 3 trials"
> 
>         pairwise_mse = []
>         for a, b in combinations(trials, 2):
>             pairwise_mse.append(mse(a, b))
> 
>         image_variabilities.append(np.mean(pairwise_mse))
> 
>     return np.mean(image_variabilities)
> 
> 
> def compute_correlations(subject_accuracies, subject_variabilities):
>     """
>     Compute Pearson and Spearman correlation.
> 
>     Args:
>         subject_accuracies: list[float]
>         subject_variabilities: list[float]
> 
>     Returns:
>         dict
>     """
>     pearson_corr, pearson_p = pearsonr(subject_accuracies, subject_variabilities)
>     spearman_corr, spearman_p = spearmanr(subject_accuracies, subject_variabilities)
> 
>     return {
>         "pearson_r": pearson_corr,
>         "pearson_p": pearson_p,
>         "spearman_r": spearman_corr,
>         "spearman_p": spearman_p,
>     }
> 
> 
> # =========================
> # Example Usage
> # =========================
> 
> num_subjects = 10
> 
> subject_accuracies = []
> subject_variabilities = []
> 
> for subject_id in range(num_subjects):
> 
>     # -------------------------
>     # Replace this with your real accuracy
>     # -------------------------
>     accuracy = np.random.rand()
>     subject_accuracies.append(accuracy)
> 
>     # -------------------------
>     # Replace this with your real trial data
>     # shape example: feature vector of dim 768
>     # -------------------------
>     subject_trials = {
>         f"img_{i}": [
>             np.random.randn(768),
>             np.random.randn(768),
>             np.random.randn(768),
>         ]
>         for i in range(100)
>     }
> 
>     variability = compute_subject_variability(subject_trials)
>     subject_variabilities.append(variability)
> 
> 
> results = compute_correlations(subject_accuracies, subject_variabilities)
> 
> print("Subject Accuracies:", subject_accuracies)
> print("Subject Variabilities:", subject_variabilities)
> print("\nCorrelation Results:")
> for k, v in results.items():
>     print(f"{k}: {v:.6f}")

太谢谢您了，社区不能没有您！不过我有个小小的疑问，什么是对一个图片产生的3次trial呢？我们的训练集不是有四次重复吗，正常来说是求平均后送入模型，那这三个trail分别是什么呢？

**Comment by HaitaoWuTJU** [comment]
不好意思打错了😬，是4次trial。
